File: src/main/java/org/example/Book.java
package org.example;

import java.time.LocalDate;
import java.util.LinkedList;
import java.util.Queue;

public class Book {
    private String title;
    private String author;
    private BookStatus status;
    private LocalDate dueDate;
    private Queue<String> holdQueue;
    private Borrower currentBorrower;

    private static int days = 14;

    //constructor
    public Book(String title, String author) {
        this.title = title;
        this.author = author;
        this.status = BookStatus.AVAILABLE;
        this.dueDate = null;
        this.holdQueue = new LinkedList<>();
        this.currentBorrower = null;
    }

    public BookStatus getStatus(){
        return status;
    }

    public String getTitle(){
        return title;
    }

    public String getAuthor(){
        return author;
    }

    public LocalDate getDueDate(){
        return dueDate;
    }

    public Queue<String> getHoldQueue() {
        return holdQueue;
    }

    public Borrower getCurrentBorrower() {
        return currentBorrower;
    }


    //setters
    public void setStatus(BookStatus newStatus){
        status = newStatus;
    }

    public void setDueDate(){
        dueDate = LocalDate.now().plusDays(days);
    }

    public void addHoldQueue(String username){
        holdQueue.add(username);
    }

    public void setCurrentBorrower(Borrower borrower){
        currentBorrower = borrower;
    }
}

File: src/models/Book.js
//models/Book.js

class Book {
    static BORROW_DAYS = 14;

    constructor(title, author) {
        this.title = title;
        this.author = author;
        this.status = 'AVAILABLE';
        this.dueDate = null;
        this.holdQueue = [];
        this.currentBorrower = null;
    }

    // getters
    getStatus() {
        return this.status;
    }

    getTitle() {
        return this.title;
    }

    getAuthor() {
        return this.author;
    }

    getDueDate() {
        return this.dueDate;
    }

    getHoldQueue() {
        return this.holdQueue;
    }

    getCurrentBorrower() {
        return this.currentBorrower;
    }

    // setters / actions
    setStatus(newStatus) {
        this.status = newStatus;
    }

    setDueDate() {
        const now = new Date();
        now.setDate(now.getDate() + Book.BORROW_DAYS);
        this.dueDate = now;
    }

    clearDueDate() {
        this.dueDate = null;
    }

    addHoldQueue(username) {
        if (!this.holdQueue.includes(username)) {
            this.holdQueue.push(username);
        }
    }

    removeBorrowerHoldQueue(username) {
        this.holdQueue = this.holdQueue.filter(u => u !== username);
    }

    setCurrentBorrower(borrower) {
        this.currentBorrower = borrower;
    }
}

module.exports = Book;